### ISSUE TITLE
Site verification fails with CiviCRM sites on D8/D9 due to folder structure differences

### ISSUE THREAD

**Issue by VangelisP** [omega8cc/boa#1592]
Assuming that we register a platform that includes CiviCRM on a D8 distro with the following structure:
```
- platform name
-- vendor
-- web
```
currently, all verifications will fails because the module `hosting_civicrm` is looking for Civicrm inside the folder pointed by `d()->root` (drupal root which is `/web`), so it searches on `/platform_name/web/vendor/civicrm/civicrm-core` which of course fails.

Borrowing the code from [this repo](https://gitlab.com/aegir/hosting_civicrm/-/blob/7.x-3.x/drush/provision_civicrm.inc#L255), I believe the best way to move forward is to convert [this part of the code](https://github.com/omega8cc/hosting_civicrm/blob/7.x-3.x/drush/provision_civicrm.inc#L283-L288) into this:

```
    if (drush_drupal_major_version() >= 8) {
      d()->root = _provision_civicrm_auto_fix_platform_root(d()->root);
      // Avoid destroying the actual drupal root folder by assigning the value into a new variable
      $root = d()->root;
      $root = preg_replace('#web/?$#', '', $root);
      $crmpath = $root . '/vendor/civicrm/civicrm-core/';

      if (is_dir($crmpath)) {
        drush_log(dt('CiviCRM: civicrm is in @path', array('@path' => $crmpath)));
        return $crmpath;
      }
    }
```

I did test it with both CiviCRM on D8 and CiviCRM on D7 and it works properly, verification succeeds.

Any thoughts?

**Comment by omega8cc** [comment]
Hmm.. that would be backwards, actually, because `_provision_civicrm_auto_fix_platform_root()` is doing exactly that in our fork to support three popular web dirs to make it better than the original code which supports only `platform_name/web` and it worked in many tests as documented with screenshots elsewhere in this issue queue.

**Comment by VangelisP** [comment]
Right, I understand, however this is weird. I actually did 2 tests:

- Registered a platform as `platform_name` (skipping the `/web`)
- Registered a platform as `platform_name/web` (including  `/web`)

Then created 2 sites, one on each platform

In both cases, while verifying, when the call hits the function `_provision_civicrm_auto_fix_platform_root()` , the `$isroot` variable always returns `/platform_name/web`, which is true and makes sense, because Drupal actually lives inside `/platform_name/web` but then none of the [conditional actions](https://github.com/omega8cc/hosting_civicrm/blob/7.x-3.x/drush/provision_civicrm.inc#L253-L263) apply at all, I see no debug (or even if i convert them to `info`) messages at all! It's like it is completely skipping ignoring the conditional check.  I do get the `Platform path needs correction, running auto-discovery..` which means we're reaching that point, but nothing after that.

Just to state that we do not run composer update/install on the environment, we usually create it locally and untar the tarball to our Aegir instance so the 'Deployment strategy' to register the platform is the " Classic (managed by you on SSH command line or SFTP) "

> it worked in many tests as documented with screenshots elsewhere in this issue queue.

I am trying to find the related issues but i am unsuccessful so far, any hints/pointers? I would appreciate it.

Many thanks !

**Comment by omega8cc** [comment]
I think that we have confused two different goals of the same procedure in Aegir own `function provision_auto_fix_platform_root($isroot)` where its sole task is to allow the user to type either `/platform_name/web` or `/platform_name/` and always make Aegir finding the correct web root path, whereas in the CiviCRM extension it should be about finding and using the raw-root with any known web dir path removed from it, so it can use it to find `/vendor/civicrm/civicrm-core/` -- I have checked the tests mentioned and realised that all were about testing different CiviCRM versions and not different Drupal directories variants.

**Comment by VangelisP** [comment]
Hi guys. This commit that you did looks like it's the proper way to move forward, especially after separating the roles of the 2 functions: `_provision_civicrm_find_platform_app_root` & `_provision_civicrm_find_platform_web_root` and enreaching each function for multiple conditions.

However, on the `_provision_civicrm_find_platform_app_root`, the `$isroot`, that is coming from `d()->root` is already being corrected/fixed to `/web` (in both my case scenarios that I've described on the OP) so the condition that applies is just this:

`$test_root_index = $isroot . '/index.php';`

which of course is wrong. I would expect the `$isroot` to point to the originally registered platform path so that the 3rd conditional action would kick in (`elseif (provision_file()->exists($test_root_web)->status())`) , where `$test_root_web = $isroot . '/web/index.php';`

TLDR: On both cases, this var `$test_root_web = $isroot . '/web/index.php';` is never being read/evaluated because there's no `/platform_name/web/web/index.php`, even though when registering the platform we've used the `platform_name` only excluding the `/web`.

Is there any object/var that we can use to return the originally registered platform path intact ? Like `d()->platform` (I don't know if it works, just saying).

Does it make any sense?

**Comment by VangelisP** [comment]
Assuming that we got no variable to bring us back the path to the platform, I've reworked the conditional actions on [this PR](https://github.com/omega8cc/hosting_civicrm/pull/1)

**Comment by omega8cc** [comment]
This looks like good and quick workaround, indeed, since the actual problem is that our Aegir Provision fork already overrides `d()->root` with lines like this:

```
d()->root = provision_auto_fix_platform_root(d()->root);
```
Which simply means we can't rely on `d()->root` not being already "fixed" in the CiviCRM extension.

Thanks!

**Comment by VangelisP** [comment]
Hi @omega8cc , there was a forgotten switch-case statement [here](https://github.com/omega8cc/hosting_civicrm/blob/7.x-3.x/drush/provision_civicrm.inc#L741-L746) that caused the break on D9.
Specifically, I got the error:
`Error: Call to undefined function db_query() in _provision_civicrm_database_version() (line 742 of /data/disk/xxx/aegir/distro/002/profiles/hostmaster/modules/aegir/hosting_civicrm/drush/provision_civicrm.inc)`

I've got a minor fix/PR for D9 that now passes verification. [PR is here](https://github.com/omega8cc/hosting_civicrm/pull/3)

**Comment by VangelisP** [comment]
Closing issue as resolved
